src/app.jsx

```diff
@@ -30,35 +29,51 @@ import { NetworkList } from "./components/networks/networkList.jsx";
 import { VmDetailsPage } from './components/vm/vmDetailsPage.jsx';
 import { CreateVmAction } from "./components/create-vm-dialog/createVmDialog.jsx";
 import LibvirtSlate from "./components/libvirtSlate.jsx";
-import { isObjectEmpty, dummyVmsFilter, vmId } from "./helpers.js";
+import { dummyVmsFilter, vmId } from "./helpers.js";
 import { InlineNotification } from 'cockpit-components-inline-notification.jsx';
 import {
+    getApiData,
     usageStartPolling,
     usageStopPolling,
 } from "./libvirt-dbus.js";
+import { getOsInfoList, getLoggedInUser } from './libvirt-common.js';
 import { useObject, useEvent } from "hooks";
 import * as service from 'service.js';
+import store from './store.js';
 
-import {
-    updateLibvirtState,
-} from './actions/store-actions.js';
+import { updateLibvirtState } from './actions/store-actions.js';
 const _ = cockpit.gettext;
 
 superuser.reload_page_on_change();
 
-export const App = ({ store, name }) => {
+export const App = ({ name }) => {
     const libvirtService = useObject(() => service.proxy(name), null, []);
+    const prevLibvirtServiceRef = useRef();
+    const [loadingResources, setLoadingResources] = useState(false);
 
     useEvent(libvirtService, "changed", () => {
         store.dispatch(updateLibvirtState({
-            activeState: libvirtService.exists ? libvirtService.state : 'stopped',
+            activeState: libvirtService.state,
         }));
+        if (libvirtService.exists && libvirtService.state == 'running' && prevLibvirtServiceRef.current !== 'running' && prevLibvirtServiceRef.current !== null) {
+            setLoadingResources(true);
+            getApiData({ connectionName: 'system' })
+                    .catch(exc => console.warn("getApiData failed: ", exc.message))
+                    .finally(() => setLoadingResources(false));
+        }
+        prevLibvirtServiceRef.current = libvirtService.state;
     });
     useEvent(superuser, "changed");
+    useEffect(() => {
+        getOsInfoList();
```

Comment:
That looks rather expensive -- useEffect() runs on every state change, no? It's fine to do this once libvirtd.service moves to "running", but e.g. not if it stops, or even if unrelated states like `loadingResources` change. You can filter for events in the second argument, but I'm not very familiar with this. But ISTM that the OS info list sohuld rather be moved into line 59, i.e. the code that runs exactly when libvirtd.service moves to "running"?

Reply:
This useEffect code runs one time when the component mounts. see [] as the dependency list. https://medium.com/@felippenardi/how-to-do-componentdidmount-with-react-hooks-553ba39d1571

getOsInfoList is not libvirt API, but libosinfo, and running that on app mount is correct I feel. It also does not need superuser to I feel there is not need to recall it on libvirtd changes.

Reply:
Sorry about that -- I misremembered what the second `[]` argument does. Thanks!